File: app/src/main/java/cn/ifafu/ifafu/mvp/other/WebActivity.java
package cn.ifafu.ifafu.mvp.other;

import android.content.Context;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.view.KeyEvent;
import android.view.View;
import android.view.ViewConfiguration;
import android.webkit.CookieManager;
import android.webkit.WebResourceRequest;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.Toolbar;

import java.net.CookieStore;
import java.net.HttpCookie;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

import cn.ifafu.ifafu.R;
import cn.woolsen.android.mvp.BaseActivity;

public class WebActivity extends BaseActivity {

    private WebView webView;

    private String loadUrl;

    private Boolean js;

    private String referer;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_web);

        webView = findViewById(R.id.webView);
        webView.setWebViewClient(new WebViewClient() {

            @Override
            public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
                view.loadUrl(request.getUrl().toString());
                return true;
            }

            @Override
            public void onPageStarted(WebView view, String url, Bitmap favicon) {
                super.onPageStarted(view, url, favicon);
            }

            @Override
            public void onPageFinished(WebView view, String url) {
                super.onPageFinished(view, url);
            }
        });

        if (referer != null) {
            Map<String, String> extraHeaders = new HashMap<>();
            extraHeaders.put("Referer", referer);
            webView.loadUrl(loadUrl, extraHeaders);
        } else {
            webView.loadUrl(loadUrl);
        }
        getStartUpParams();
        setWebViewSetting();
    }

    @Override
    protected void onDestroy() {
        if (webView != null) {
            webView.setVisibility(View.GONE);
            long timeout = ViewConfiguration.getZoomControlsTimeout();
            webView.postDelayed(() -> webView.destroy(), timeout);
        }
        super.onDestroy();
    }

    private void setWebViewSetting() {
        WebSettings webSettings = webView.getSettings();

        webSettings.setJavaScriptEnabled(js);
        webSettings.setUseWideViewPort(true);
        webSettings.setLoadWithOverviewMode(true);
        webSettings.setSupportZoom(true);
        webSettings.setBuiltInZoomControls(true);
        webSettings.setLoadsImagesAutomatically(true);
        webSettings.setDomStorageEnabled(true);
        webSettings.setAppCacheEnabled(true);
        webSettings.setAppCachePath(
                getApplicationContext().getDir("cache", Context.MODE_PRIVATE).getPath());
    }

    private void getStartUpParams() {
        Bundle bundle = getIntent().getExtras();
        loadUrl = bundle.getString("loadUrl");

        Toolbar toolbar = findViewById(R.id.tb_web);
        toolbar.setTitle(bundle.getString("pageTitle"));
        toolbar.setNavigationOnClickListener(v -> back());

        js = bundle.getBoolean("js", true);

        if (bundle.getBoolean("cookie", false)) {
            CookieManager cookieManager = CookieManager.getInstance();

            CookieStore cookieStore = new java.net.CookieManager().getCookieStore();

            List<HttpCookie> cookies = cookieStore.getCookies();
            for (HttpCookie cookie : cookies) {
                cookieManager.setCookie(cookie.getDomain(), String.format(
                        Locale.getDefault(), "%s; domain=%s; path=%s",
                        cookie.toString(), cookie.getDomain(), cookie.getPath()));
            }
        }

        referer = bundle.getString("referer", null);
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_BACK && event.getAction() == KeyEvent.ACTION_DOWN) {
            back();
            return true;
        }

        return super.onKeyDown(keyCode, event);
    }

    private void back() {
        if (webView.canGoBack()) {
            webView.goBack();
        } else {
            webView.destroy();
            finish();
        }
    }

}

File: app/src/main/java/cn/ifafu/ifafu/ui/web/WebActivity.kt
package cn.ifafu.ifafu.ui.web

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.net.Uri
import android.os.Bundle
import android.provider.MediaStore
import android.view.KeyEvent
import android.webkit.*
import android.widget.LinearLayout
import androidx.activity.viewModels
import androidx.lifecycle.Observer
import cn.ifafu.ifafu.R
import cn.ifafu.ifafu.ui.getViewModelFactory
import cn.ifafu.ifafu.base.BaseActivity
import cn.ifafu.ifafu.databinding.ActivityWebBinding
import cn.ifafu.ifafu.ui.view.LoadingDialog
import kotlinx.android.synthetic.main.activity_web.*

class WebActivity : BaseActivity() {

    private val mLoadingDialog = LoadingDialog(this, "加载中")

    private lateinit var webView: WebView

    private var mFilePathCallback: ValueCallback<Array<Uri>>? = null
    private val PHOTO_REQUEST_CODE = 1023

    private var first = true

    private val mViewModel: WebViewModel by viewModels { getViewModelFactory() }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setLightUiBar()
        bind<ActivityWebBinding>(R.layout.activity_web)
        initWebView()
        mViewModel.loadUrl.observe(this, Observer {
            webView.loadUrl(it)
        })
        mViewModel.title.observe(this, Observer {
            tb_web.title = it
        })
        mViewModel.init(intent)
    }


    @SuppressLint("SetJavaScriptEnabled")
    private fun initWebView() {
        webView = WebView(this)
        view_root.addView(webView, LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.MATCH_PARENT))
        webView.webViewClient = object : WebViewClient() {
            override fun onPageFinished(view: WebView?, url: String?) {
                super.onPageFinished(view, url)
                mLoadingDialog.cancel()
            }

            override fun onPageStarted(view: WebView?, url: String?, favicon: Bitmap?) {
                super.onPageStarted(view, url, favicon)
                mLoadingDialog.show()
            }

            override fun shouldOverrideUrlLoading(view: WebView?, request: WebResourceRequest?): Boolean {
                if (first && intent.hasExtra("referer")) {
                    val referer = intent.getStringExtra("referer")
                    request?.requestHeaders?.put("referer", referer)
                    first = false
                }
                return super.shouldOverrideUrlLoading(view, request)
            }
        }

        webView.webChromeClient = object : WebChromeClient() {
            override fun onShowFileChooser(webView: WebView?, filePathCallback: ValueCallback<Array<Uri>>?, fileChooserParams: FileChooserParams?): Boolean {
                mFilePathCallback = filePathCallback
                val intent = Intent(Intent.ACTION_PICK, MediaStore.Images.Media.EXTERNAL_CONTENT_URI)
                startActivityForResult(intent, PHOTO_REQUEST_CODE)
                return true
            }
        }
        val webSettings = webView.settings
        webSettings.javaScriptEnabled = true
        webSettings.useWideViewPort = true
        webSettings.loadWithOverviewMode = true
        webSettings.setSupportZoom(true)
        webSettings.builtInZoomControls = true
        webSettings.loadsImagesAutomatically = true
        webSettings.domStorageEnabled = true
        webSettings.setAppCacheEnabled(true)
        val cachePath = applicationContext.getDir("cache", Context.MODE_PRIVATE).path
        webSettings.setAppCachePath(cachePath)
        btn_refresh.setOnClickListener { webView.reload() }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == PHOTO_REQUEST_CODE && resultCode == Activity.RESULT_OK) {
            data?.data?.run {
                mFilePathCallback?.onReceiveValue(arrayOf(this))
            }
        }
        super.onActivityResult(requestCode, resultCode, data)
    }


    override fun onDestroy() {
        if (mLoadingDialog.isShowing) {
            mLoadingDialog.cancel()
        }
        super.onDestroy()
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        if (event.action == KeyEvent.ACTION_DOWN) {
            //按返回键操作并且能回退网页
            if (keyCode == KeyEvent.KEYCODE_BACK && webView.canGoBack()) {
                //后退
                webView.goBack()
                return true
            }
        }
        return super.onKeyDown(keyCode, event)
    }
}
